src/geometry/ArborX_GeometryTraits.hpp

```diff
@@ -0,0 +1,102 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef ARBORX_GEOMETRY_TRAITS_HPP
+#define ARBORX_GEOMETRY_TRAITS_HPP
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX
+{
+
+namespace GeometryTraits
+{
+
+struct PointTag
+{};
+
+struct BoxTag
+{};
+
+struct SphereTag
+{};
+
+template <typename Geometry>
+struct dimension
+{
+  using not_specialized = void; // tag to detect existence of a specialization
+};
+
+template <typename Geometry>
+struct tag
+{
+  using not_specialized = void; // tag to detect existence of a specialization
+};
+
+template <typename Geometry>
+using DimensionNotSpecializedArchetypeAlias =
+    typename dimension<Geometry>::not_specialized;
+
+template <typename Geometry>
+using TagNotSpecializedArchetypeAlias = typename tag<Geometry>::not_specialized;
+
+template <typename Geometry>
+using DimensionArchetypeAlias = decltype(dimension<Geometry>::value);
+
+template <typename Geometry>
+using TagArchetypeAlias = typename tag<Geometry>::type;
+
+template <typename Geometry>
+struct is_point
+    : Kokkos::is_detected_exact<PointTag, TagArchetypeAlias, Geometry>
+{};
```

Comment:
Why not just
```suggestion
template <typename Geometry>
struct is_point
    : std::is_same<typename tag<Geometry>::type, PointTag>::value
{};
```
and similar below without introducing all these `*Archtype*` aliases?

I could see the detection idiom to be more useful if we were to require the user's classes to have the respective aliases.

Reply:
We still need archetypes for validating the provided traits. I don't have a strong a strong opinion about the proposed patch. Just feel like introducing the archetypes allows us to specify the signature in a single place, so that we don't need to worry about specific details in other places. For example, if we rename the member alias from `type` to something else, we would need to change it in multiple places if not using archetypes.

Reply:
For me, seeing `is_detected` always carries a big cognitive load (especially since it will not be standardized) and I find it easier to read "regular" type traits if possible instead.